File: html/class_page.rb
require 'interface_database.rb',
        'interface_list_item.rb',
        'page.rb',
        'search_page.rb',
        'string_utils.rb' do

class ClassPage < Page

  attr_accessor :interface
  attr_accessor :current_member

  PLACEMARKER_UPDATE_INTERVAL = 0.1

  def initialize(url: nil)
    super('class')

    # FIXME: Redirect to search when a class isn't found or if there is no class.
    # Where do we do this? In something like view_did_appear?

    if url
      class_name, member_name = url.split('/')
      if class_name
        self.interface = InterfaceDatabase.find_interface(name: class_name, type: :class)
        self.current_member = find_member(member_name)
      end
    end
  end

  def self.is_valid_url(url)
    return false if !url

    class_name, _ = url.split('/')
    interface = InterfaceDatabase.find_interface(name: class_name, type: :class)
    return false if !interface

    true
  end

  def interface=(new_interface)
    @interface = new_interface

    if @interface
      @methods = @interface[:methods]
      @attributes = @interface[:attributes]
      parent = @interface[:parent]
      while parent
        @methods += parent[:methods]
        @attributes += parent[:attributes]
        parent = parent[:parent]
      end
      @methods.sort_by! { |method| method[:name] }
      @attributes.sort_by! { |attribute| attribute[:name] }
    end

    update_interface_elements if @element
  end

  def current_member=(new_member)
    @current_member = new_member
    WebDocs.page_stack.update_location_bar_url

    update_current_member_dimming
  end

  def location_bar_url
    return "/#{@page_name}" if !@interface
    return "/#{@page_name}/#{interface[:name]}" if !@current_member
    member_name = Documentation.underscore(@current_member[:name])
    return "/#{@page_name}/#{interface[:name]}/#{member_name}"
  end

  def will_appear
    $window.add_event_listener('popstate', method(:on_pop_state))
    update_current_member_dimming
  end

  def did_appear
    update_placemarker
  end

  def will_disappear
    $window.remove_event_listener('popstate', method(:on_pop_state))
  end

  protected

  def did_load
    @content_container = @element.query_selector('.content')
    @methods_list = @element.query_selector('.member-list.methods')
    @attributes_list = @element.query_selector('.member-list.attributes')
    @attributes_sidebar_list = @element.query_selector('.sidebar-list.attributes')
    @methods_sidebar_list = @element.query_selector('.sidebar-list.methods')
    @title_element = @element.query_selector('header>h1')
    @interface_description_element = @element.query_selector('.interface-description')
    @interface_list_item = InterfaceListItem.new
    @interface_list_item.is_header_clickable = false
    @interface_description_element.append_child(@interface_list_item)
    @placemarker = @element.query_selector('.place-marker')
    @sidebar_list_container = @element.query_selector('.sidebar-list-container')

    @content_container.onscroll = method(:on_scroll_content)
    @element.query_selector('.search-button').onclick = method(:on_click_search)

    update_interface_elements
    scroll_to_member(@current_member)
  end

  private

  def update_current_member_dimming
    return if !@attributes_list || !@methods_list
    all_list_items = @attributes_list.child_nodes.entries + @methods_list.child_nodes.entries
    if !@current_member
      all_list_items.each { |item| item.class_list.remove('dim') }
      $window.set_timeout(300) do
        all_list_items.each { |item| item.class_list.remove('transition') }
      end
    else
      member_index = (@attributes + @methods).index(@current_member)      
      all_list_items.each_with_index do |item, index|
        if index == member_index
          item.class_list.remove('dim', 'transition')
        else
          item.class_list.add('dim', 'transition')
        end
      end
    end
  end

  def on_click_search(event)
    search_page = SearchPage.new
    WebDocs.page_stack.push(page: search_page, animated:true)
  end

  def on_scroll_content(event)
    if !@scrolling_programmatically && @current_member
      self.current_member = nil
    end
    @scrolling_programmatically = false

    update_placemarker
  end

  def update_placemarker
    # Don't update the placemarker position too often. Otherwise, it's slow.
    now = Time.now.to_f
    time_since_last_update = now - (@last_placemarker_update || 0)
    return if time_since_last_update < PLACEMARKER_UPDATE_INTERVAL
    @last_placemarker_update = now.to_f

    # Find the top and bottom visible elements, then move the marker to cover those.
    minimum_top = @content_container.scroll_top
    maximum_bottom = @content_container.scroll_top + @content_container.offset_height
    all_children = @attributes_list.child_nodes.entries + @methods_list.child_nodes.entries
    all_sidebar_children = @attributes_sidebar_list.child_nodes.entries + @methods_sidebar_list.child_nodes.entries
    visible = all_children.find_all do |child|
      (child.offset_top + child.offset_height > minimum_top) && (child.offset_top < maximum_bottom)
    end

    first_visible_index = all_children.index(visible.first)
    last_visible_index = all_children.index(visible.last)
    first_sidebar_item = all_sidebar_children[first_visible_index]
    last_sidebar_item = all_sidebar_children[last_visible_index]

    first_sidebar_top = first_sidebar_item.offset_top + 1
    last_sidebar_bottom = last_sidebar_item.offset_top + last_sidebar_item.offset_height - 2
    height = last_sidebar_bottom - first_sidebar_top

    @placemarker.style.top = "#{first_sidebar_top}px"
    @placemarker.style.height = "#{height}px"
  end

  def update_interface_elements
    @attributes_list.inner_html = ''
    @methods_list.inner_html = ''
    @attributes_sidebar_list.inner_html = ''
    @methods_sidebar_list.inner_html = ''
    @title_element.inner_html = ''
    @interface_list_item.interface = nil

    return if !@interface

    @element.query_selector_all('.attributes').each do |attributes_element|
      attributes_element.style.display = @attributes.length == 0 ? 'none' : 'block'
    end

    @element.query_selector_all('.methods').each do |methods_element|
      methods_element.style.display = @methods.length == 0 ? 'none' : 'block'
    end

    @title_element.inner_text = @interface[:name]
    @interface_list_item.interface = @interface

    @methods.each { |method| add_member_elements(method, @methods_list, @methods_sidebar_list) }
    @attributes.each { |attribute| add_member_elements(attribute, @attributes_list, @attributes_sidebar_list) }
  end

  def add_member_elements(member, list, sidebar_list)
    list_item = InterfaceListItem.new
    list_item.interface = member
    show_class = member[:owner_id] != @interface[:id]
    list_item.is_header_clickable = show_class
    list_item.show_owner_class = show_class
    list_item.add_event_listener(InterfaceListItem::CLICKED_INTERFACE, method(:on_click_member))
    list.append_child(list_item)

    sidebar_item = $window.document.create_element('li')
    sidebar_item.inner_text = Documentation.underscore(member[:name])
    sidebar_item.class_list.add(member[:interface_type].to_s)
    sidebar_item.class_list.add('ellipsize')
    sidebar_list.append_child(sidebar_item)

    sidebar_item.onclick do
      scroll_to_member(member)

      return if self.current_member == member
      self.current_member = member
    end
  end

  def on_click_member(event)
    member = event.detail.interface
    class_page = ClassPage.new
    class_page.interface = member[:owner]
    class_page.current_member = member
    WebDocs.page_stack.push(page: class_page, animated: true)
  end

  def scroll_to_member(member)
    return if !member

    case member[:interface_type]
    when :attribute
      data_array = @attributes
      list = @attributes_list
    when :method
      data_array = @methods
      list = @methods_list
    end

    index = data_array.index(member)
    return if !index

    element = list.child_nodes[index]
    return if !element
    center_element_on_page(element)
  end

  def center_element_on_page(element)
    item_rect = element.get_bounding_client_rect

    # Wait until the page is loaded
    if (item_rect.width == 0 || item_rect.height == 0)
      $window.set_timeout(0) { center_element_on_page(element) }
      return
    end

    target_y = @content_container.scroll_top + item_rect.top + (item_rect.height / 2)
    new_scroll_y = target_y - (@content_container.client_height / 2)
    @scrolling_programmatically = true
    @content_container.scroll_top = new_scroll_y
  end

  def find_member(member_name)
    return nil if !@attributes || !@methods || !member_name
    lowercase_member_name = Documentation.lower_camel_case(member_name).downcase
    (@attributes + @methods).find { |member| member[:name].downcase == lowercase_member_name }
  end

  def on_pop_state(event)
    $window.console.log("ClassPage on_pop_state", event.state)
    return if event.state[:class_page] != object_id

    $window.console.log("ClassPage on_pop_state for this page.")
    member = find_member(event.state[:member_name])
    scroll_to_member(member)
  end

end # SearchController

end # require

File: site/scripts/class-page.js
var ClassPage = function(url) {
  Page.call(this, 'class');

  if (url) {
    urlParts = url.split('/');
    var className = urlParts[0];
    var memberName = urlParts[1];
    if (className) {
      this.interfaceData = InterfaceDatabase.findInterface(className, 'class');
      this.currentMember = this._findMember(memberName);
    }
  }
};

ClassPage.prototype = new Page();

ClassPage.prototype.PLACEMARKER_UPDATE_INTERVAL = 0.1;

Object.defineProperty(ClassPage.prototype, 'interfaceData', {
  get: function() {
    return this._interfaceData;
  },
  set: function(newData) {
    this._interfaceData = newData;

    if (this._interfaceData) {
      this._methods = this._interfaceData.methods;
      this._attributes = this._interfaceData.attributes;
      var parent = this._interfaceData.parent;
      while (parent) {
        this._methods = this._methods.concat(parent.methods);
        this._attributes = this._attributes.concat(parent.attributes);
        parent = parent.parent;
      }
      var stringSort = function(a, b) {
        return a.name.localeCompare(b.name);
      };
      this._methods.sort(stringSort);
      this._attributes.sort(stringSort);
    }

    if (this.element) {
      this._updateInterfaceElements();
    }
  }
});


Object.defineProperty(ClassPage.prototype, 'currentMember', {
  get: function() {
    return this._currentMember;
  },
  set: function(newMember) {
    this._currentMember = newMember;
    WebDocs.pageStack.updateLocationBarUrl();
    this._updateCurrentMemberDimming();
  }
});

ClassPage.isValidUrl = function(url) {
  if (!url) {
    return false;
  }

  var className = url.split('/')[0];
  var interfaceData = InterfaceDatabase.findInterface(className, 'class');
  if (!interfaceData) {
    return false;
  }

  return true;
};

ClassPage.prototype.locationBarUrl = function() {
  var url = '/' + this.pageName;
  if (this._interfaceData) {
    url += '/' + this._interfaceData.name;
  }
  if (this._currentMember) {
    url += '/' + this._currentMember.name;
  }
  return url;
}

ClassPage.prototype.pageTitle = function() {
  if (!this._interfaceData) {
    return null;
  }

  return this._interfaceData.name + ': EasyDocs';
};

ClassPage.prototype._willAppear = function() {
  this._updateCurrentMemberDimming();
};

ClassPage.prototype._didAppear = function() {
  this._updatePlacemarker();
};

ClassPage.prototype._didLoad = function() {
  this._contentContainer = this.element.querySelector('.content');
  this._methodsList = this.element.querySelector('.member-list.methods');
  this._attributesList = this.element.querySelector('.member-list.attributes');
  this._attributesSidebarList = this.element.querySelector('.sidebar-list.attributes');
  this._methodsSidebarList = this.element.querySelector('.sidebar-list.methods');
  this._titleElement = this.element.querySelector('header>h1');
  this._interfaceDescriptionElement = this.element.querySelector('.interface-description');
  this._interfaceListItem = new InterfaceListItem();
  this._interfaceListItem.isHeaderClickable = false;
  this._interfaceDescriptionElement.appendChild(this._interfaceListItem.element);
  this._placemarker = this.element.querySelector('.place-marker');
  this._sidebarListContainer = this.element.querySelector('.sidebar-list-container');

  this._contentContainer.onscroll = this._onScrollContent;
  this.element.querySelector('.search-button').onclick = this._onClickSearch;

  this._updateInterfaceElements();
  this._scrollToMember(this.currentMember);
};

ClassPage.prototype._updateCurrentMemberDimming = function() {
  if (!this._attributesList || !this._methodsList) {
    return;
  }

  var attributeItems = Array.prototype.slice.call(this._attributesList.childNodes);
  var methodItems = Array.prototype.slice.call(this._methodsList.childNodes);
  var allListItems = attributeItems.concat(methodItems);
  if (!this.currentMember) {
    this._undimAll();
  } else {
    this._dimAllExceptCurrentMember();
  }
};

ClassPage.prototype._undimAll = function() {
  allListItems.forEach(function(item) {
    item.classList.remove('dim');
  });
  window.setTimeout(function() {
    allListItems.forEach(function(item) {
      item.classList.remove('transition');
    });
  }, 1);
};

ClassPage.prototype._dimAllExceptCurrentMember = function() {
  var allMembers = this._attributes.concat(this._methods);
  var memberIndex = allMembers.index(this._currentMember);
  allListItems.forEach(function(item, index) {
    if (index == memberIndex) {
      item.classList.remove('dim', 'transition');
    } else {
      item.classList.add('dim', 'transition');
    }
  });
};

ClassPage.prototype._onClickSearch = function(event) {
  var searchPage = new SearchPage();
  WebDocs.pageStack.push(searchPage, true);
};

ClassPage.prototype._onScrollContent = function(event) {
  if (!this._scrollingProgrammatically && this._currentMember) {
    this.currentMember = null;
  }
  this._scrollingProgrammatically = false;

  this._updatePlacemarker();
};

ClassPage.prototype._updatePlacemarker = function() {
  // // Don't update the placemarker position too often. Otherwise, it's slow.
  // now = Time.now.to_f
  // time_since_last_update = now - (@last_placemarker_update || 0)
  // return if time_since_last_update < PLACEMARKER_UPDATE_INTERVAL
  // @last_placemarker_update = now.to_f

  // Find the top and bottom visible elements, then move the marker to cover those.
  var minimumTop = this._contentContainer.scrollTop;
  var maximumBottom = this._contentContainer.scrollTop + this._contentContainer.offsetHeight;
  var allChildren = nodeListToArray(this._attributesList.childNodes).concat(nodeListToArray(this._methodsList.childNodes));
  var visibleChildren = allChildren.filter(function(child) {
    return (child.offsetTop + child.offsetHeight > minimumTop) && (child.offsetTop < maximumBottom);
  });

  var firstVisibleIndex = allChildren.indexOf(visibleChildren[0]);
  var lastVisibleIndex = allChildren.index(visibleChildren[visibleChildren.length - 1]);

  var allSidebarChildren = nodeListToArray(this._attributesSidebarList.childNodes).concat(nodeListToArray(this._methodsSidebarList.childNodes));
  var firstSidebarItem = allSidebarChildren[firstVisibleIndex];
  var lastSidebarItem = allSidebarChildren[lastVisibleIndex];

  var firstSidebarTop = firstSidebarItem.offsetTop + 1;
  var lastSidebarBottom = lastSidebarItem.offsetTop + lastSidebarItem.offsetHeight - 2;
  var height = lastSidebarBottom - firstSidebarTop;

  this._placemarker.style.top = firstSidebarTop + 'px';
  this._placemarker.style.height = height + 'px';
};

ClassPage.prototype._updateInterfaceElements = function() {
  this._attributesList.innerHTML = '';
  this._methodsList.innerHTML = '';
  this._attributesSidebarList.innerHTML = '';
  this._methodsSidebarList.innerHTML = '';
  this._titleElement.innerHTML = '';
  this._interfaceListItem.interface = null;

  if (!this.interfaceData) {
    return;
  }

  this.element.querySelectorAll('.attributes').forEach(function(attributesElement) {
    attributesElement.style.display = this._attributes.length === 0 ? 'none' : 'block';
  });

  this.element.querySelectorAll('.methods').forEach(function(methodsElement) {
    methodsElement.style.display = this._methods.length === 0 ? 'none' : 'block';
  });

  this._titleElement.innerText = this.interfaceData.name;
  this._interfaceListItem.interfaceData = this.interfaceData;

  this._methods.forEach(function(method) {
    this._addMemberElements(method, this._methodsList, this._methodsSidebarList);
  }, this);
  this._attributes.forEach(function(attribute) {
    this._addMemberElements(attribute, this._attributesList, this._attributesSidebarList);
  }, this);
};

ClassPage.prototype._addMemberElements = function(member, list, sidebarList) {
  var listItem = new InterfaceListItem();
  listItem.interfaceData = member;
  var showClass = (member.ownerId != this.interfaceData.id);
  listItem.isHeaderClickable = showClass;
  listItem.showOwnerClass = showClass;
  listItem.addEventListener(InterfaceListItem.CLICKED_INTERFACE, this._onClickMember);
  list.appendChild(listItem.element);

  var sidebarItem = document.createElement('li');
  sidebarItem.innerText = member.name;
  sidebarItem.classList.add(member.interfaceType);
  sidebarItem.classList.add('ellipsize');
  sidebarList.appendChild(sidebarItem);

  sidebarItem.onclick = function() {
    this._scrollToMember(member);

    if (self.currentMember === member) {
      return;
    }
    self.currentMember = member;
  };
};

ClassPage.prototype._onClickMember = function(event) {
  member = event.detail.interfaceData;
  var classPage = new ClassPage();
  classPage.interfaceData = member.owner;
  classPage.currentMember = member;
  WebDocs.pageStack.push(classPage, true);
};

ClassPage.prototype._scrollToMember = function(member) {
  if (!member) {
    return;
  }

  var dataArray, list;
  switch (member.interfaceType) {
  case 'attribute':
    dataArray = this._attributes;
    list = this._attributesList;
    break;
  case 'method':
    dataArray = this._methods;
    list = this._methodsList;
  }

  var index = dataArray.indexOf(member);
  if (index === -1) {
    return;
  }

  var element = list.childNodes.item(index);
  if (!element) {
    return;
  }

  this._centerElementOnPage(element);
};

ClassPage.prototype._centerElementOnPage = function(element) {
  var itemRect = element.getBoundingClientRect();

  // Wait until the page is loaded
  if (itemRect.width === 0 || itemRect.height === 0) {
    window.setTimeout(function() {
      this._centerElementOnPage(element);
    }, 0);
    return;
  }

  var targetY = this._contentContainer.scrollTop + itemRect.top + (itemRect.height / 2);
  var newScrollY = targetY - (this._contentContainer.clientHeight / 2);
  this._scrollingProgrammatically = true;
  this._contentContainer.scrollTop = newScrollY;
};

ClassPage.prototype._findMember = function(memberName) {
  if (!this._attributes || !this._methods || !memberName) {
    return null;
  }

  var lowercaseMemberName = memberName.toLowerCase();
  var allMembers = this._attributes.concat(this._methods);
  allMembers.forEach(function(member) {
    if (member.name.toLowerCase() == lowercaseMemberName) {
      return member;
    }
  });

  return null;
};